File: token/token.go
package token

import "fmt"

type TokenType string

const (
	ILLEGAL = "ILLEGAL"
	EOF     = "EOF"
	// identifiers and literals
	IDENTIFIER = "IDENTIFIER"
	INTEGER    = "INTEGER"
	// operators
	ASSIGNMENT   = "="
	PLUS         = "+"
	MINUS        = "-"
	BANG         = "!"
	ASTERISK     = "*"
	SLASH        = "/"
	LESS_THAN    = "<"
	GREATER_THAN = ">"
	EQUALS       = "=="
	NOT_EQUALS   = "!="
	// delimiters
	COMMA     = ","
	PAREN_L   = "("
	PAREN_R   = ")"
	BRACE_L   = "{"
	BRACE_R   = "}"
	SEMICOLON = ";"
	//keywords
	FUNCTION    = "FUNCTION"
	DECLARATION = "DECLARATION"
	TRUE        = "TRUE"
	FALSE       = "FALSE"
	IF          = "IF"
	ELSE        = "ELSE"
	RETURN      = "RETURN"
)

type Token struct {
	Type    TokenType
	Literal string
	Pos     Pos
	Len     int
}

func (t Token) String() string {
	typeAndLiteral := string(t.Type)
	if typeAndLiteral != t.Literal {
		typeAndLiteral += " " + t.Literal
	}
	return fmt.Sprintf("{%s %v}", typeAndLiteral, t.Pos)
}

type Pos struct {
	Ln  int
	Col int
}

func (p Pos) String() string {
	return fmt.Sprintf("%d:%d", p.Ln, p.Col)
}

var keywords = map[string]TokenType{
	"fn":     FUNCTION,
	"let":    DECLARATION,
	"true":   TRUE,
	"false":  FALSE,
	"if":     IF,
	"else":   ELSE,
	"return": RETURN,
}

func LookupIdentifier(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return IDENTIFIER
}

File: src/token/token.zig
const std = @import("std");

pub const Token = struct {
    type: TokenType,
    literal: []const u8,
    pos: Pos,

    pub fn string(
        self: Token,
        allocator: std.mem.Allocator,
    ) []const u8 {
        return std.fmt.allocPrint(
            allocator,
            "[{s}: '{s}']",
            .{ @tagName(self.type), self.literal },
        ) catch |err| @errorName(err);
    }
};

pub const Pos = struct {
    ln: u32,
    col: u32,

    pub fn string(
        self: Pos,
        allocator: std.mem.Allocator,
    ) []const u8 {
        return std.fmt.allocPrint(
            allocator,
            "{d}:{d}",
            .{ self.ln, self.col },
        ) catch |err| @errorName(err);
    }
};

pub const TokenType = enum {
    illegal,
    eof,

    // identifiers and literals
    identifier,
    integer,
    string,

    // operators
    assign,
    plus,
    minus,
    asterisk,
    slash,
    bang,
    less_than,
    greater_than,
    equals,
    not_equals,

    // delimiters
    comma,
    colon,
    semicolon,
    paren_open,
    paren_close,
    brace_open,
    brace_close,
    bracket_open,
    bracket_close,

    // keywords
    function,
    declaration,
    true,
    false,
    @"if",
    @"else",
    @"return",

    pub fn string(
        self: TokenType,
        allocator: std.mem.Allocator,
    ) []const u8 {
        return std.fmt.allocPrint(
            allocator,
            "[{s}]",
            .{@tagName(self)},
        ) catch |err| @errorName(err);
    }
};

const Keyword = enum {
    @"fn",
    let,
    true,
    false,
    @"if",
    @"else",
    @"return",
};

pub fn lookupIdentifier(identifier: []const u8) TokenType {
    if (std.meta.stringToEnum(Keyword, identifier)) |keyword| {
        return switch (keyword) {
            .@"fn" => .function,
            .let => .declaration,
            .true => .true,
            .false => .false,
            .@"if" => .@"if",
            .@"else" => .@"else",
            .@"return" => .@"return",
        };
    } else {
        return .identifier;
    }
}
